### ISSUE TITLE
skiplist synchronization question

### ISSUE THREAD

**Issue by zzt93** [google/leveldb#428]
The comment in the skiplist.h says like following:
```
void SkipList<Key,Comparator>::Insert(const Key& key) {
  // TODO(opt): We can use a barrier-free variant of FindGreaterOrEqual()
  // here since Insert() is externally synchronized.
```
But when I following the source code, I found no external synchronization in db_impl.cc:DBImpl::Write.
```
        // extract sequence and kType to insert memtable
        status = WriteBatchInternal::InsertInto(updates, mem_);
```
Is this comment out of date or I missed something?

**Comment by ghemawat** [comment]
On Thu, Nov 17, 2016 at 2:59 AM, Tony Zeng <notifications@github.com> wrote:

> The comment in the skiplist.h says like following:
>
> void SkipList<Key,Comparator>::Insert(const Key& key) {
>   // TODO(opt): We can use a barrier-free variant of FindGreaterOrEqual()
>   // here since Insert() is externally synchronized.
>
> But when I following the source code, I found no external synchronization
> in db_impl.cc:DBImpl::Write.
>
>         // extract sequence and kType to insert memtable
>         status = WriteBatchInternal::InsertInto(updates, mem_);
>
> Is this comment out of date or I missed something?
>
See the other comment in Write:
    // Add to log and apply to memtable.  We can release the lock
    // during this phase since &w is currently responsible for logging
    // and protects against concurrent loggers and concurrent writes

The earlier code has already ensured that this thread is the one
responsible for writing to log.

**Comment by zzt93** [comment]
Yes, earlier code has ensured no other thread will interrupt this writer/logger, but without lock and memory barrier version of `FindGreaterOrEqual()`, will reader thread always read the newest value that writer thread add?

**Comment by ghemawat** [comment]
The comment this bug references is for a writer-writer race, which does not
exist as the comment and my message says.

W.r.t.. a reader always reading the newest value, that is guaranteed by the
SetNext() at the bottom of Insert which does a release store.  Insert's
call to FindGreaterOrEqual does not matter to a reader since
FindGreaterOrEqual does not make any changes.

**Comment by zzt93** [comment]
Thank you. I understand now.
One more question: Why not just give up a giant mutex lock and use atomic variable to solve writer-writer race? shouldn't atomic operation be cheaper?
